pkg/spdk/engine.go

```diff
@@ -1289,48 +1300,56 @@ func (e *Engine) replicaShallowCopy(srcReplicaServiceCli, dstReplicaServiceCli *
 		}
 	}()
 
+	e.log.Infof("Engine is starting snapshots shallow copy from rebuilding src replica %s to rebuilding dst replica %s", srcReplicaName, dstReplicaName)
+
 	rebuildingSnapshotMap := map[string]*api.Lvol{}
 	for _, snapshotApiLvol := range rebuildingSnapshotList {
 		rebuildingSnapshotMap[snapshotApiLvol.Name] = snapshotApiLvol
 	}
 
 	// Traverse the src replica snapshot tree with a DFS way and do shallow copy one by one
-	currentSnapshotName, prevSnapshotName := "", ""
+	timer := time.NewTimer(MaxShallowCopyWaitTime)
+	defer timer.Stop()
+	ticker := time.NewTicker(ShallowCopyCheckInterval)
+	defer ticker.Stop()
+	currentSnapshotName := ""
 	for idx := 0; idx < len(rebuildingSnapshotList); idx++ {
 		currentSnapshotName = rebuildingSnapshotList[idx].Name
 		e.log.Debugf("Engine is syncing snapshot %s from rebuilding src replica %s to rebuilding dst replica %s", currentSnapshotName, srcReplicaName, dstReplicaName)
-		// TODO: Handle backing image
-		if prevSnapshotName == "" || rebuildingSnapshotMap[currentSnapshotName].Parent != prevSnapshotName {
-			if err = srcReplicaServiceCli.ReplicaRebuildingSrcDetach(srcReplicaName, dstReplicaName); err != nil {
-				return err
-			}
-			// Create or Recreate a rebuilding lvol behinds the parent of the current snapshot
-			dstRebuildingLvolAddress, err := dstReplicaServiceCli.ReplicaRebuildingDstSnapshotRevert(dstReplicaName, rebuildingSnapshotMap[currentSnapshotName].Parent)
-			if err != nil {
-				return err
-			}
-			if err = srcReplicaServiceCli.ReplicaRebuildingSrcAttach(srcReplicaName, dstReplicaName, dstRebuildingLvolAddress); err != nil {
-				return err
-			}
-		}
 
 		if err := dstReplicaServiceCli.ReplicaRebuildingDstShallowCopyStart(dstReplicaName, currentSnapshotName); err != nil {
 			return errors.Wrapf(err, "failed to start shallow copy snapshot %s", currentSnapshotName)
 		}
-		for {
-			shallowCopyStatus, err := dstReplicaServiceCli.ReplicaRebuildingDstShallowCopyCheck(dstReplicaName)
-			if err != nil {
-				return err
-			}
-			if shallowCopyStatus.State == helpertypes.ShallowCopyStateError || shallowCopyStatus.Error != "" {
-				return fmt.Errorf("rebuilding error during shallow copy for snapshot %s: %s", shallowCopyStatus.SnapshotName, shallowCopyStatus.Error)
-			}
-			if shallowCopyStatus.State == helpertypes.ShallowCopyStateComplete {
-				if shallowCopyStatus.Progress != 100 {
-					e.log.Warnf("Shallow copy snapshot %s is %s but somehow the progress is not 100%%", shallowCopyStatus.SnapshotName, helpertypes.ShallowCopyStateComplete)
+
+		timer.Reset(MaxShallowCopyWaitTime)
+		continuousRetryCount := 0
+		for finished := false; !finished; {
```

Comment:
Can we use 
```
for {
    select {
    case <-timer.C:
        ...
    case <-ticker.C:
        ...
        if shallowCopyStatus.State == helpertypes.ShallowCopyStateComplete {
            ...
            break
        }
    }
}
```
and remove `finished` flag?

Reply:
NO.  The `break` in your code means leaving from `select` rather than the whole `for` loop